WebKit doesn't build with clang stable/21.x branch

rniwa opened the pull request:
#### a24facd1f7d97b1895f5732489014d346034eb8f
<pre>
WebKit doesn&apos;t build with clang stable/21.x branch
<a href="https://bugs.webkit.org/show_bug.cgi?id=298854">https://bugs.webkit.org/show_bug.cgi?id=298854</a>

Reviewed by Chris Dumez.

This PR fixes a bunch of compiler warnings in WebKit emitted by clang stable/21.x branch.

No new tests since there should be no behavioral change.

* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig: Disable nontrivial-memcall and
uninitialized-const-pointer warnings in ANGLE as they result in a compiler error.
* Source/WebCore/Modules/mediastream/RTCRtpTransceiver.h: Made the destructor non-virtual
since this class is final.
* Source/WebCore/Modules/webaudio/DynamicsCompressorNode.h: Ditto for a member function.
* Source/WebCore/css/CSSFontFace.h: Ditto.
* Source/WebCore/html/track/VideoTrack.h: Ditto.
* Source/WebCore/loader/icon/IconLoader.cpp:
(WebCore::IconLoader::notifyFinished): Use unsafeMakeSpan instead of std::span constructor
since the latter will result in a warning.
* Source/WebCore/platform/graphics/GeometryUtilities.cpp:
(WebCore::toRelatedAcuteAngle): Simplified the code since angle &gt; 90 || angle &lt; 180 is
always true.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.h:
Removed the unused member variable.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
Made a member function non-virtual since this class is final.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp: Removed the unused
member variable.
* Source/WebCore/rendering/RenderListBox.h: Made a member function non-virtual since this
class is final.
* Source/WebCore/svg/SVGPathByteStreamBuilder.h:
(WebCore::SVGPathByteStreamBuilder::writeType): Use asByteSpan instead of std::span
constructor since the latter results a warning.
* Source/WebCore/svg/SVGStyleElement.h: Made a member function non-virtual since this
class is final.
* Source/WebCore/xml/XMLHttpRequest.h: Ditto.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp: Removed an unused member variable.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp: Ditto.
* Source/WebGPU/WebGPU/CommandEncoder.h: Ditto.
* Source/WebGPU/WebGPU/WebGPU.h:
(wgpuGetComputePassDescriptorTimestampWrites): Replaced __attribute__((lifetimebound))
annotation with LIFETIME_BOUND macro and moved it to after the parameter name as this
attribute only applies there. Also dropped noescape attribute since the parameter does
escape the function via its return value.
(wgpuGetRenderPassDescriptorTimestampWrites): Ditto.
(wgpuGetRenderPassDescriptorColorAttachments): Ditto.
(wgpuGetRenderPassDescriptorDepthSencilAttachment): Ditto.
* Source/WebKit/GPUProcess/webrtc/UserMediaCaptureManagerProxy.cpp: Removed an unused
member variable.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h: Made the
destructor non-virtual since this class is final.

Canonical link: <a href="https://commits.webkit.org/300006@main">https://commits.webkit.org/300006@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/801eaff8294e721063f28e65b204a618346c5422

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120977 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40673 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31331 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127394 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73057 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/02d324ae-4ea0-4312-9b6a-82da65d9d301) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122853 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49252 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91889 "10 flakes 57 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61128 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3408ff95-15b0-4a5c-91cb-f9aea88a0879) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123929 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33039 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108453 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72577 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/72a8cc9f-287e-4d8e-b8aa-9be82c454e85) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32065 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26555 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70983 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102542 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26735 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130249 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47904 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36398 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100497 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104620 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100399 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45808 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23855 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44592 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19202 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47762 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53475 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47233 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50580 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48917 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/platform/graphics/GeometryUtilities.cpp:
```diff
@@ -265,9 +265,8 @@ float toRelatedAcuteAngle(float angle)
-        return std::abs(180 - angle);
-    return std::abs(360 - angle);
+    // FIXME: We likely don't handle an angle between 180 and 360 degrees correctly.
+    return std::abs(180 - angle);
```
Doesn't this change the path for an angle that is exactly 90? Before it would return 270 and now 90.

mwyrzykowski replied:
➕

rniwa replied:
No, if angle = 90, then angle < 180. The condition is always true.

mwyrzykowski replied:
should we fix the intent here in that case? Likely it should have been:
```
if (angle < 270)
```
without the preceding `||`?

rniwa replied:
That sounds like something a person more familiar with this code should do in a separate PR. I tried a bunch of values of `ray()` but couldn't make it behave differently from Chrome.

mwyrzykowski replied:
that would creating the mapping:
```
angle < 90: f(angle) => angle ∈ [0, 90]
90 <= angle <= 270: f(angle) => |angle - 180| ∈ [0, 90]
270 <= angle  <= 360: f(angle) => 360 - angle ∈ [0, 90]
```
these are all acute angles

ddkilzer replied:
We need to write a unit test for this.

Maybe remove this from this PR and file a follow-up bug?

mwyrzykowski replied:
That would probably be better. The FIXME without a Bugzilla is likely to go unnoticed like this issue since it was committed 3 years ago

rniwa replied:
ok. Added a FIXME referring to webkit.org/b/298890.

ddkilzer commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h:
```diff
@@ -107,10 +107,10 @@ struct BackgroundLayer {
     PREFERRED_TYPE(FillAttachment) unsigned m_attachment : 2;
     PREFERRED_TYPE(FillBox) unsigned m_clip : 3;
-    PREFERRED_TYPE(FillBox) unsigned m_origin : 2;
+    PREFERRED_TYPE(FillBox) unsigned m_origin : 3;
```
This should use `WebCore::FillLayer::FillBoxBitWidth` from `Source/WebCore/rendering/style/FillLayer.h`.

ddkilzer replied:
Make `WebCore::FillLayer::FillBoxBitWidth` public if necessary.

rniwa replied:
I don't see that constant defined anywhere? In fact, `Source/WebCore/rendering/style/FillLayer.h` doesn't even exist.

ddkilzer commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h:
```diff
@@ -107,10 +107,10 @@ struct BackgroundLayer {
     PREFERRED_TYPE(BlendMode) unsigned m_blendMode : 5;
 
-    PREFERRED_TYPE(FillBox) mutable unsigned m_clipMax : 2; // maximum m_clip value from this to bottom layer
+    PREFERRED_TYPE(FillBox) mutable unsigned m_clipMax : 3; // maximum m_clip value from this to bottom layer
```
This should use `WebCore::FillLayer::FillBoxBitWidth` from `Source/WebCore/rendering/style/FillLayer.h`.

ddkilzer commented on Source/WebCore/style/values/masking/StyleMaskLayer.h:
```diff
@@ -110,11 +110,11 @@ struct MaskLayer {
 
     PREFERRED_TYPE(FillBox) unsigned m_clip : 3;
-    PREFERRED_TYPE(FillBox) unsigned m_origin : 2;
+    PREFERRED_TYPE(FillBox) unsigned m_origin : 3;
```
This should use `WebCore::FillLayer::FillBoxBitWidth` from `Source/WebCore/rendering/style/FillLayer.h`.

ddkilzer commented on Source/WebCore/style/values/masking/StyleMaskLayer.h:
```diff
@@ -110,11 +110,11 @@ struct MaskLayer {
     PREFERRED_TYPE(MaskMode) unsigned m_maskMode : 2;
 
-    PREFERRED_TYPE(FillBox) mutable unsigned m_clipMax : 2; // maximum m_clip value from this to bottom layer
+    PREFERRED_TYPE(FillBox) mutable unsigned m_clipMax : 3; // maximum m_clip value from this to bottom layer
```
This should use `WebCore::FillLayer::FillBoxBitWidth` from `Source/WebCore/rendering/style/FillLayer.h`.

mwyrzykowski reviewed (dismissed):
I don’t know why the behavior is changed as @annevk commented

ddkilzer commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1201,7 +1201,7 @@ typedef struct WGPUComputePassDescriptor {
 } SWIFT_ESCAPABLE WGPUComputePassDescriptor WGPU_STRUCTURE_ATTRIBUTE;
 
-static inline WGPU_NULLABLE WGPUComputePassTimestampWrites const * _Nullable __counted_by(1) wgpuGetComputePassDescriptorTimestampWrites(const WGPUComputePassDescriptor * __attribute__((lifetimebound)) __counted_by(1) __attribute__((noescape)) descriptor) {
+static inline WGPU_NULLABLE WGPUComputePassTimestampWrites const * _Nullable __counted_by(1) wgpuGetComputePassDescriptorTimestampWrites(const WGPUComputePassDescriptor * __counted_by(1) __attribute__((noescape)) descriptor) {
```
Seems like `__attribute__((noescape))` is what should be removed here, not `__attribute__((lifetimebound))`.

The function returns `descriptor->timestampWrites`, which means a value from `descriptor` escapes.

rniwa replied:
That doesn't work:
```
/Volumes/Data/safari-4/OpenSource/Source/WebGPU/WebGPU/WebGPU.h:1204:187: error: 'lifetimebound' attribute only applies to parameters and implicit object parameters
 1204 | static inline WGPU_NULLABLE WGPUComputePassTimestampWrites const * _Nullable __counted_by(1) wgpuGetComputePassDescriptorTimestampWrites(const WGPUComputePassDescriptor * __attribute__((lifetimebound)) __counted_by(1) descriptor) {
      |                                                                                                                                                                                           ^
/Volumes/Data/safari-4/OpenSource/Source/WebGPU/WebGPU/WebGPU.h:1337:163: error: 'lifetimebound' attribute only applies to parameters and implicit object parameters
 1337 | inline WGPURenderPassTimestampWrites const * _Nullable __counted_by(1) wgpuGetRenderPassDescriptorTimestampWrites(const WGPURenderPassDescriptor * __attribute__((lifetimebound)) __counted_by(1) descriptor) {
      |                                                                                                                                                                   ^
/Volumes/Data/safari-4/OpenSource/Source/WebGPU/WebGPU/WebGPU.h:1341:158: error: 'lifetimebound' attribute only applies to parameters and implicit object parameters
 1341 | inline WGPURenderPassColorAttachment const * __counted_by(count) wgpuGetRenderPassDescriptorColorAttachments(const WGPURenderPassDescriptor * __attribute__((lifetimebound)) __counted_by(1) descriptor, size_t count) {
      |                                                                                                                                                              ^
/Volumes/Data/safari-4/OpenSource/Source/WebGPU/WebGPU/WebGPU.h:1346:176: error: 'lifetimebound' attribute only applies to parameters and implicit object parameters
 1346 | inline WGPURenderPassDepthStencilAttachment const * _Nullable __counted_by(1) wgpuGetRenderPassDescriptorDepthSencilAttachment(const WGPURenderPassDescriptor * __attribute__((lifetimebound)) __counted_by(1) descriptor) {
      |                                                                                                                                                                                ^
4 errors generated.
```

mwyrzykowski replied:
maybe @geoffreygaren has an idea?

ddkilzer commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1334,16 +1334,16 @@ typedef struct WGPURenderPassDescriptor {
 } SWIFT_ESCAPABLE WGPURenderPassDescriptor WGPU_STRUCTURE_ATTRIBUTE;
 
-inline WGPURenderPassTimestampWrites const * _Nullable __counted_by(1) wgpuGetRenderPassDescriptorTimestampWrites(const WGPURenderPassDescriptor * __attribute__((lifetimebound)) __counted_by(1) __attribute__((noescape)) descriptor) {
+inline WGPURenderPassTimestampWrites const * _Nullable __counted_by(1) wgpuGetRenderPassDescriptorTimestampWrites(const WGPURenderPassDescriptor * __counted_by(1) __attribute__((noescape)) descriptor) {
```
Seems like `__attribute__((noescape))` is what should be removed here, not `__attribute__((lifetimebound))`.

The function returns `descriptor->timestampWrites`, which means a value from `descriptor` escapes.

ddkilzer commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1334,16 +1334,16 @@ typedef struct WGPURenderPassDescriptor {
 }
 
-inline WGPURenderPassColorAttachment const * __counted_by(count) wgpuGetRenderPassDescriptorColorAttachments(const WGPURenderPassDescriptor * __attribute__((lifetimebound)) __counted_by(1) __attribute__((noescape)) descriptor, size_t count) {
+inline WGPURenderPassColorAttachment const * __counted_by(count) wgpuGetRenderPassDescriptorColorAttachments(const WGPURenderPassDescriptor * __counted_by(1) __attribute__((noescape)) descriptor, size_t count) {
```
Seems like `__attribute__((noescape))` is what should be removed here, not `__attribute__((lifetimebound))`.

The function returns `descriptor->colorAttachments`, which means a value from `descriptor` escapes.

ddkilzer commented on Source/WebGPU/WebGPU/WebGPU.h:
```diff
@@ -1334,16 +1334,16 @@ typedef struct WGPURenderPassDescriptor {
 }
 
-inline WGPURenderPassDepthStencilAttachment const * _Nullable __counted_by(1) wgpuGetRenderPassDescriptorDepthSencilAttachment(const WGPURenderPassDescriptor * __attribute__((lifetimebound)) __counted_by(1) __attribute__((noescape)) descriptor) {
+inline WGPURenderPassDepthStencilAttachment const * _Nullable __counted_by(1) wgpuGetRenderPassDescriptorDepthSencilAttachment(const WGPURenderPassDescriptor * __counted_by(1) __attribute__((noescape)) descriptor) {
```
Seems like `__attribute__((noescape))` is what should be removed here, not `__attribute__((lifetimebound))`.

The function returns `descriptor->depthStencilAttachment`, which means a value from `descriptor` escapes.

ddkilzer reviewed (changes requested):
Please use `FillBoxBitWidth` and re-review the `lifetimebounds` attribute removals.

(Sorry, I went to the gym between starting and finishing the review, so others might have noted these already.)

rniwa committed:
WebKit doesn't build with clang stable/21.x branch
https://bugs.webkit.org/show_bug.cgi?id=298854

Reviewed by Chris Dumez.

This PR fixes a bunch of compiler warnings in WebKit emitted by clang stable/21.x branch.

No new tests since there should be no behavioral change.

* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig: Disable nontrivial-memcall and
uninitialized-const-pointer warnings in ANGLE as they result in a compiler error.
* Source/WebCore/Modules/mediastream/RTCRtpTransceiver.h: Made the destructor non-virtual
since this class is final.
* Source/WebCore/Modules/webaudio/DynamicsCompressorNode.h: Ditto for a member function.
* Source/WebCore/css/CSSFontFace.h: Ditto.
* Source/WebCore/html/track/VideoTrack.h: Ditto.
* Source/WebCore/loader/icon/IconLoader.cpp:
(WebCore::IconLoader::notifyFinished): Use unsafeMakeSpan instead of std::span constructor
since the latter will result in a warning.
* Source/WebCore/platform/graphics/GeometryUtilities.cpp:
(WebCore::toRelatedAcuteAngle): Simplified the code since angle > 90 || angle < 180 is
always true.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.h:
Removed the unused member variable.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
Made a member function non-virtual since this class is final.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp: Removed the unused
member variable.
* Source/WebCore/rendering/RenderListBox.h: Made a member function non-virtual since this
class is final.
* Source/WebCore/svg/SVGPathByteStreamBuilder.h:
(WebCore::SVGPathByteStreamBuilder::writeType): Use asByteSpan instead of std::span
constructor since the latter results a warning.
* Source/WebCore/svg/SVGStyleElement.h: Made a member function non-virtual since this
class is final.
* Source/WebCore/xml/XMLHttpRequest.h: Ditto.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp: Removed an unused member variable.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp: Ditto.
* Source/WebGPU/WebGPU/CommandEncoder.h: Ditto.
* Source/WebGPU/WebGPU/WebGPU.h:
(wgpuGetComputePassDescriptorTimestampWrites): Replaced __attribute__((lifetimebound))
annotation with LIFETIME_BOUND macro and moved it to after the parameter name as this
attribute only applies there. Also dropped noescape attribute since the parameter does
escape the function via its return value.
(wgpuGetRenderPassDescriptorTimestampWrites): Ditto.
(wgpuGetRenderPassDescriptorColorAttachments): Ditto.
(wgpuGetRenderPassDescriptorDepthSencilAttachment): Ditto.
* Source/WebKit/GPUProcess/webrtc/UserMediaCaptureManagerProxy.cpp: Removed an unused
member variable.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h: Made the
destructor non-virtual since this class is final.

Canonical link: https://commits.webkit.org/300006@main
